sg_start.8
.TH SG_START "8" "October 2006" "sg3_utils-1.22" SG3_UTILS
.SH NAME
sg_start \- starts (spins-up) or stops (spins down) device, load or
eject medium
.SH SYNOPSIS
.B sg_start
[\fI--eject\fR] [\fI--fl=<n>\fR] [\fI-i\fR] [\fI--imm=0|1\fR] [\fI--load\fR]
[\fI--loej\fR] [\fI--pc=<n>\fR] [\fI--start\fR] [\fI--stop\fR] [\fI-v\fR]
[\fI-V\fR] [\fI0|1\fR] \fI<device>\fR
.SH DESCRIPTION
.\" Add any additional description here
.PP
sg_start sends a START STOP UNIT SCSI command to the given device with
the selected options. The most used options are '--stop' to spin down a disk
and '--start' to spin up a disk. Using 'sg_start --start /dev/sda' on a disk
that is already spinning is harmless. There is also finer grain control
with "power conditions": active, idle and standby. In some contexts
the "stop" state can be considered an additional power condition. 
.PP
Devices that contain removable media such as cd/dvds can use the '--loej'
flag to load the medium when used in conjunction with '--start' (i.e. load
medium then spin up). Alternatively '--loej' may be used to eject the medium
when used alone or when used in conjunction with '--stop' (i.e. spin down
then eject medium). More simply the loading and ejecting of a removable 
medium can be requested with the '--load' and '--eject' options.
.PP
If no option or argument is given then a '--start' is assumed; as the
utility's name suggests.
.TP
0
stop (spin-down) given device.
Active when '--pc=" is not given or '--pc=0'.
.TP
1
start (spin-up) given device.
Active when '--pc=" is not given or '--pc=0'. Harmless if device has
already started (or is in the process of starting).
.TP
--eject
stop the medium and eject it from the drive. Only appropriate for a
device with removable medium. Might be ignored (prevented), see below.
.TP
--fl=<n>
sets the format layer number for the disc to "jump" to (defined in
MMC-5). Values of '<n>' can be 0 to 3. When this option is chosen,
the FL, LoEj and Start bits are set in the cdb as required by MMC-5; thus
the user does not need to set the '--start' and/or '--load' options. 
.TP
-i
sets the IMM bit on the START STOP UNIT command so this utility will
return immediately and not wait for the media to spin down. Same
effect as '--imm=1'. The default action (without this option or
a '--imm=1' option) is to wait until the media spins down before
returning.
.TP
--imm=0|1
when the immediate bit is 1 then this utility returns immediately after
the device has received the command. When this option is 0 (the default)
then the utility returns once the command has completed its action
(i.e. it waits until the device is started or stopped).
.TP
--load
load the medium in the drive and start it. Only appropriate for a
removable medium.
.TP
--loej
load the media when the unit is started or eject it when the unit is
stopped. This option is ignored if 'pc > 0'. Default is off (i.e. don't
attempt to load or eject media). If a start/start indication is not
given (neither "--start" nor "--stop") and this option is given then a
stop and eject action is assumed.
.TP
--pc=<n>
set the 'power conditions' value (in hex); 0 to f (inclusive) are valid.
Default value is 0.
When '--pc=0' then '--eject', '--load', '--loej', '--start' and '--stop'
are active. Some common values are 1 for the active power condition (SBC);
2 for the idle power condition; 3 for the standby power condition; 5 for
sleep power condition (MMC); 7 for LU_CONTROL (SBC), 0xa (decimal 10) for
FORCE_IDLE_0 (SBC) and 0xb (decimal 11) for FORCE_STANDBY_0 (SBC).
See recent SBC-3, MMC-5 and SAS drafts at www.t10.org for more information.
.TP
--start
start (spin-up) given device. Using this option on an already started
device is harmless. Same meaning as "1" argument.
.TP
--stop
stop (spin-down) given device. Same meaning as "0" argument.
.TP
-v
verbose: outputs SCSI command in hex to console before with executing
it. '-vv' and '-vvv' are also accepted yielding greater verbosity.
.TP
-V
print out version string then exit.
.PP
All options, other than '-v' and '-V', can be given with a single "-".
For example: "sg_start -stop /dev/sda" and "sg_start --stop /dev/sda"
are equivalent. The single "-" form is for backward compatibility.
.PP
To avoid confusion, only one of "0", "1", "--load", "--eject", "--start"
and "--stop" can be given.
.PP
There is an associated "power condition" mode page (0x1a) in which timer
values can be set for transitioning to either idle or standby state after
a period of inactivity. The sdparm utility can be used to view the
power condition mode page and if required change it.
If a device is in either idle or standby power condition state then
a REQUEST SENSE command (see the sg_requests utility) should yield 
a sense key of "no sense" and an additional sense code of "Low
power condition on" on recent SCSI devices.
.PP
Ejection of removable media (e.g. 'sg_start --eject /dev/hdd' where
the device is an ATAPI cd/dvd drive) may be prevented by a prior
PREVENT ALLOW MEDIUM REMOVAL SCSI command (see sg_prevent). In this
case this utility should fail with an error generated by the device:
illegal request / medium removal prevented. This can be overridden
using sg_prevent or 'sdparm --command=unlock /dev/hdd'.
.PP
The SCSI TEST UNIT READY command can be used to find out whether a
device is ready to transfer data. If rotating media is stopped or
still coming up to speed, then the TEST UNIT READY command will yield
a "not ready" sense key and an more informative additional sense
code. See the sg_turs utility.
.PP
In the 2.4 series of Linux kernels the given device must be
a SCSI generic (sg) device. In the 2.6 series block devices (e.g. disks
and SCSI DVDs) can also be specified. For example "sg_start 0 /dev/sda"
will work in the 2.6 series kernels.
.SH NOTES
Earlier versions had a '-s' option to perform a SYNCHRONIZE CACHE command
before the START STOP UNIT command was issued. According to recent SBC-2
drafts this is done implicitly if required. Hence the "-s" option has been
dropped.
.SH EXIT STATUS
The exit status of sg_start is 0 when it is successful. Otherwise see
the sg3_utils(8) man page.
.SH AUTHOR
Written by K. Garloff and D. Gilbert
.SH "REPORTING BUGS"
Report bugs to <dgilbert at interlog dot com>.
.SH COPYRIGHT
Copyright \(co 2002-2006 Kurt Garloff, Douglas Gilbert
.br
This software is distributed under the GPL version 2. There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH "SEE ALSO"
.B sg_prevent(sg3_utils), sg_requests(sg3_utils), sg_turs(sg3_utils)
.B sdparm(sdparm)


CHANGELOG
Each utility has its own version number, date of last change and
some description at the top of its ".c" file. All utilities in the main
directory have their own "man" pages. There is also a sg3_utils man page.

Changelog for sg3_utils-1.25 [20070629]
  - sg_dd: add oflag=sparse to step over bs*bpt number of zeros;
    - with oflag=sparse, write last bs*bpt segment at end or after
      error so file length of OFILE is appropriate
    - allow iflag=null and oflag=null which do nothing (placeholders)
    - when coe>1 then SCSI READ LONG logic remember extended block
      length of first encountered error
  - sg_ses: sync with ses2r17

Changelog for sg3_utils-1.24 [20070507]
  - sg_raw: new utility to send arbirary SCSI commands
  - sg_luns: increase number of luns that can be fetched
    - fix length of raw and hex output
    - add '--quiet' option to output only ASCII hex
      representation of each lun
  - sg_rtpg: update for changes in spc4r09
  - sg_persist: update documentation, spc-4 references
    - fix exit status values
  - sg_inq: update version descriptors per spc4r09
    - fix '--id' and '--extended'
    - extend block limits VPD page (sbc3r09)
  - sg_vpd: extend block limits VPD page (sbc3r09)
    - append relative target port identifier to SAS target
      port address with '-iq' option
  - sg_logs: add decoding for stats+performance log pages
    - fix showing of page names for pdt > 0
    - implement '-HH' for single and all pages, fix '-r'
    - when '--maxlen=' given, only do single fetch
    - add Tape Alert (ssc), Media and Element statistics (smc) pages 
    - add '--brief' option
  - sg_ses: sync with ses2r16
    - fix bay number for SAS
  - sg_format: add '--dcrt' and '--security' options
  - sgm_dd: add 'smmap' oflag for shared_mmap_io testing
    - add 'dio' oflag
  - sg_dd, sgp_dd: add 'dio' iflag and oflag
  - sg_modes: change SAS mode page names per sas2r09
    - check validity of block descriptors length
  - sg_pt: change opaque context object from 'void *'
    to 'struct sg_pt_base *'
  - sg_opcodes: anticipate extra tmfs from 07-159r0
  - sg_sat_set_features: add more usage information
    - add man page
  - sg_sat_phy_event: add to examples directory
  - sg_lib: sync asc/ascq codes with spc4r10
  - Solaris port: using uscsi interface

Changelog for sg3_utils-1.23 [20070131]
  - sg_read_buffer: new utility
  - sg_write_buffer: new utility
  - sg_opcodes, sg_senddiag, sg_logs, sg_modes, sg_start, sg_inq,
    sg_turs, sg_readcap, sg_rbuf: add getopt_long() based cli;
    old and new cli selectable, new getopt_long cli is default
  - scripts: new subdirectory containing some bash scripts
    - add scripts/README file
  - sg_reassign: add '--hex' option for grown and primary lists
  - sg_rtpg: add '--raw' option
  - sg_lib.h, sg_cmds_basic.h + sg_cmds_extra.h: add C++ 
    'extern "C" ' wrappers
    - cleanup C code so it will compile as C++
  - sg_lib: sync with spc4r08
    - include <inttypes.h>, use PRId64 instead of %lld form
    - fix sg_get_sense_str() when empty sense buffer
  - win32 port: add Makefile.mingw + related support for MinGW
  - sg_cmds_extra: add sg_ll_read_buffer() and sg_ll_write_buffer()
  - sg_dd, sgp_dd, sgm_dd, sg_read: use lseek64() instead of llseek.c
  - sgm_dd: accept coe=<n> for interworking with sg_dd
  - sg_rdac: fix on non-linux ports
  - sg_ses: fix spurious warning in additional element status page
    - '-rr' option outputs a diagnostic page in binary to stdout
  - sg_opcodes: add command timeout descriptor support (spc4r08)
    - change linux specific pass through to generic pass through
  - sg_logs: add 'name=value' decoding for SAS specific lpage
  - examples+utils subdirectories: remove symlinks
  - synchronize man pages with usage messages
  - sg3_utils.spec: rework

Changelog for sg3_utils-1.22 [20061016]
  - sgp_dd: accept verbose=<n> as well as deb=<n> to ease
    interworking with sg_dd and sgm_dd
  - sg_sat_set_features: added to examples directory
  - sg_lib: sync asc/ascq text with spc4r06
    - move SG_LIB_CAT_NO_SENSE and SG_LIB_CAT_RECOVERED to
      20 and 21 respectively; add SG_LIB_CAT_ABORTED_COMMAND
      at 11 (its sense key value) 
  - sg_vpd: tweak '--page=sp --quiet' output
    - change '-HHH' so same as '-rr' (prepares ATA Information
      (ai) response for hdparm)
  - sg_requests: add '-s' option to set exit status from
    parameter data
  - sg_modes: exit quickly from '-e' if device not ready
  - sg_logs: sync sas log pages with sas2r05a
    - expand background scan results log page
    - add '-m=<max_len>' to limit response length
    - drop '-scum' and '-sthr' options and add '-select'
  - sg_write_long: add '--16' option to send 16 byte cdb
    - add '--wr_uncor' and '--pblock' options
  - sg_senddiag: cleanup and add sdiag_sas_p1_stop.txt
    to examples directory
  - sg_format: add '--cmplst=<n>' option (default: 1)
    - add '--pfu=<n>' option
    - expand man page to discuss P/D/C/GLISTs
  - sg_reassign: add '--primary' option to fetch primary
    defect list (PLIST) length
  - sg_readcap: add '-H' option to output response in hex
    and '-r' to output response in binary to stdout
    - add logical blocks per physical block (sbc3r07)
  - sginfo: add PLIST and GLIST designation to defect lists
  - sg_cmds: split this support file into sg_cmds_basic.[hc]
    and sg_cmds_extra.[hc]
    - add sg_ll_ata_pt() (SATL ATA pass) to sg_cmds_extra.[hc]
  - sg_rdac: fix includes for FreeBSD
  - sg_dd: add 'coe_limit=' option to exit after <n>
    consecutive 'coe' type read errors
  - sgm_dd: print out throughput information when signal arrives
    if time=1 (like sg_dd does already)
  - sg_inq: change '-HHH' so same as '-rr'. Now sg_inq, sg_vpd
    and sdparm output for hdparm with '-HHH'
    -add '-l=<resp_len>' option
  - sg_read_long: add '--pblock' option for physical blocks
  - sg_luns: add '--hex' and '--raw' options
  - sg_requests: add '--hex' and '--raw' options

sg_start.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <fcntl.h>
#include <getopt.h>

#include "sg_lib.h"
#include "sg_cmds_basic.h"

/*
 *  Copyright (C) 1999-2007 D. Gilbert
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2, or (at your option)
 *  any later version.

    Start/Stop parameter by Kurt Garloff <garloff at suse dot de>, 6/2000
    Sync cache parameter by Kurt Garloff <garloff at suse dot de>, 1/2001
    Guard block device answering sg's ioctls. 
                     <dgilbert at interlog dot com> 12/2002
    Convert to SG_IO ioctl so can use sg or block devices in 2.6.* 3/2003

    This utility was written for the Linux 2.4 kernel series. It should
    now build for the Linux 2.6 kernel series and various other Operating
    Systems.
 
*/

static char * version_str = "0.55 20070419";

static struct option long_options[] = {
        {"eject", 0, 0, 'e'},
        {"fl", 1, 0, 'f'},
        {"help", 0, 0, 'h'},
        {"immed", 0, 0, 'i'},
        {"load", 0, 0, 'l'},
        {"loej", 0, 0, 'L'},
        {"new", 0, 0, 'N'},
        {"old", 0, 0, 'O'},
        {"pc", 1, 0, 'p'},
        {"start", 0, 0, 's'},
        {"stop", 0, 0, 'S'},
        {"verbose", 0, 0, 'v'},
        {"version", 0, 0, 'V'},
        {0, 0, 0, 0},
};

struct opts_t {
    int do_eject;
    int do_fl;
    int do_help;
    int do_immed;
    int do_load;
    int do_loej;
    int do_pc;
    int do_start;
    int do_stop;
    int do_verbose;
    int do_version;
    const char * device_name;
    int opt_new;
};

void usage()
{
        fprintf(stderr, "Usage: sg_start [--eject] [--fl=FL] [--help] "
                "[--immed] [--load] [--loej]\n"
                "                [--pc=PC] [--start] [--stop] "
                "[--verbose] [--version]\n"
                "                DEVICE\n"
                "  where:\n"
                "    --eject|-e      stop unit then eject the medium\n"
                "    --fl=FL|-f FL    format layer number (mmc5)\n"
                "    --help|-h       print usage message then exit\n"
                "    --immed|-i      device should return control after "
                "receiving cdb,\n"
                "                    default action is to wait until action "
                "is complete\n"
                "    --load|-l       load medium then start the unit\n"
                "    --loej|-L       load or eject, corresponds to LOEJ bit "
                "in cdb;\n"
                "                    load when START bit also set, else "
                "eject\n"
                "    --pc=PC|-p PC    power conditions: 0 (default) -> no "
                "power condition,\n"
                "                    1 -> active, 2 -> idle, 3 -> standby, "
                "5 -> sleep (MMC)\n"
                "    --start|-s      start unit, corresponds to START bit "
                "in cdb,\n"
                "                    default (START=1) if no other options "
                "given\n"
                "    --stop|-S       stop unit (e.g. spin down disk)\n"
                "    --verbose|-v    increase verbosity\n"
                "    --version|-V    print version string then exit\n\n"
                "    Example: 'sg_start --stop /dev/sdb'    stops unit\n"
                "             'sg_start --eject /dev/scd0'  stops unit and "
                "ejects medium\n\n"
                "Performs a SCSI START STOP UNIT command\n"
                );
}

void usage_old()
{
        fprintf(stderr, "Usage:  sg_start [0] [1] [--eject] [--fl=FL] "
                "[-i] [--imm=0|1]\n"
                "                 [--load] [--loej] [--pc=PC] [--start] "
                "[--stop] [-v] [-V]\n"
                "                 DEVICE\n"
                "  where:\n"
                "    0          stop unit (e.g. spin down a disk or a "
                "cd/dvd)\n"
                "    1          start unit (e.g. spin up a disk or a "
                "cd/dvd)\n"
                "    --eject    stop then eject the medium\n"
                "    --fl=FL    format layer number (mmc5)\n"
                "    -i         return immediately (same as '--imm=1')\n"
                "    --imm=0|1  0->await completion(def), 1->return "
                "immediately\n"
                "    --load     load then start the medium\n"
                "    --loej     load the medium if '-start' option is "
                "also given\n"
                "               or stop unit and eject\n"
                "    --pc=PC    power conditions (in hex, default 0 -> no "
                "power condition)\n"
                "               1 -> active, 2 -> idle, 3 -> standby, "
                "5 -> sleep (MMC)\n"
                "    --start    start unit (same as '1'), default "
                "action\n"
                "    --stop     stop unit (same as '0')\n"
                "    -v         verbose (print out SCSI commands)\n"
                "    -V         print version string then exit\n\n"
                "    Example: 'sg_start --stop /dev/sdb'    stops unit\n"
                "             'sg_start --eject /dev/scd0'  stops unit and "
                "ejects medium\n\n"
                "Performs a SCSI START STOP UNIT command\n"
                );
}

static int process_cl_new(struct opts_t * optsp, int argc, char * argv[])
{
    int c, n, err;

    while (1) {
        int option_index = 0;

        c = getopt_long(argc, argv, "ef:hilLNOp:sSvV", long_options,
                        &option_index);
        if (c == -1)
            break;

        switch (c) {
        case 'e':
            ++optsp->do_eject;
            ++optsp->do_loej;
            break;
        case 'f':
            n = sg_get_num(optarg);
            if ((n < 0) || (n > 3)) {
                fprintf(stderr, "bad argument to '--fl='\n");
                usage();
                return SG_LIB_SYNTAX_ERROR;
            }
            ++optsp->do_loej;
            ++optsp->do_start;
            optsp->do_fl = n;
            break;
        case 'h':
        case '?':
            ++optsp->do_help;
            break;
        case 'i':
            ++optsp->do_immed;
            break;
        case 'l':
            ++optsp->do_load;
            ++optsp->do_loej;
            break;
        case 'L':
            ++optsp->do_loej;
            break;
        case 'N':
            break;      /* ignore */
        case 'O':
            optsp->opt_new = 0;
            return 0;
        case 'p':
            n = sg_get_num(optarg);
            if ((n < 0) || (n > 15)) {
                fprintf(stderr, "bad argument to '--pc='\n");
                usage();
                return SG_LIB_SYNTAX_ERROR;
            }
            optsp->do_pc = n;
            break;
        case 's':
            ++optsp->do_start;
            break;
        case 'S':
            ++optsp->do_stop;
            break;
        case 'v':
            ++optsp->do_verbose;
            break;
        case 'V':
            ++optsp->do_version;
            break;
        default:
            fprintf(stderr, "unrecognised option code %c [0x%x]\n", c, c);
            if (optsp->do_help)
                break;
            usage();
            return SG_LIB_SYNTAX_ERROR;
        }
    }
    err = 0;
    for (; optind < argc; ++optind) {
        if (1 == strlen(argv[optind])) {
            if (0 == strcmp("0", argv[optind])) {
                ++optsp->do_stop;
                continue;
            } else if (0 == strcmp("1", argv[optind])) {
                ++optsp->do_start;
                continue;
            }
        }
        if (NULL == optsp->device_name)
            optsp->device_name = argv[optind];
        else {
            fprintf(stderr, "Unexpected extra argument: %s\n", argv[optind]);
            ++err;
        }
    }
    if (err) {
        usage();
        return SG_LIB_SYNTAX_ERROR;
    } else
        return 0;
}

static int process_cl_old(struct opts_t * optsp, int argc, char * argv[])
{
    int k, jmp_out, plen, num;
    int ambigu = 0;
    int startstop = -1;
    unsigned int u;
    const char * cp;

    for (k = 1; k < argc; ++k) {
        cp = argv[k];
        plen = strlen(cp);
        if (plen <= 0)
            continue;
        if ('-' == *cp) {
            for (--plen, ++cp, jmp_out = 0; plen > 0;
                 --plen, ++cp) {
                switch (*cp) {
                case 'i':
                    if ('\0' == *(cp + 1))
                        optsp->do_immed = 1;
                    else
                        jmp_out = 1;
                    break;
                case 'v':
                    ++optsp->do_verbose;
                    break;
                case 'V':
                    ++optsp->do_version;
                    break;
                case 'h':
                case '?':
                    ++optsp->do_help;
                    break;
                case 'N':
                    optsp->opt_new = 1;
                    return 0;
                case 'O':
                    break;
                case '-':
                    ++cp;
                    --plen;
                    jmp_out = 1;
                    break;
                default:
                    jmp_out = 1;
                    break;
                }
                if (jmp_out)
                    break;
            }
            if (plen <= 0)
                continue;

            if (0 == strncmp(cp, "eject", 5)) {
                optsp->do_loej = 1;
                if (startstop == 1)
                    ambigu = 1;
                else
                    startstop = 0;
            } else if (0 == strncmp("fl=", cp, 3)) {
                num = sscanf(cp + 3, "%x", &u);
                if (1 != num) {
                    fprintf(stderr, "Bad value after 'fl=' option\n");
                    usage_old();
                    return SG_LIB_SYNTAX_ERROR;
                }
                startstop = 1;
                optsp->do_loej = 1;
                optsp->do_fl = u;
            } else if (0 == strncmp("imm=", cp, 4)) {
                num = sscanf(cp + 4, "%x", &u);
                if ((1 != num) || (u > 1)) {
                    fprintf(stderr, "Bad value after 'imm=' option\n");
                    usage_old();
                    return SG_LIB_SYNTAX_ERROR;
                }
                optsp->do_immed = u;
            } else if (0 == strncmp(cp, "load", 4)) {
                optsp->do_loej = 1;
                if (startstop == 0)
                    ambigu = 1;
                else
                    startstop = 1;
            } else if (0 == strncmp(cp, "loej", 4))
                optsp->do_loej = 1;
            else if (0 == strncmp("pc=", cp, 3)) {
                num = sscanf(cp + 3, "%x", &u);
                if ((1 != num) || (u > 15)) {
                    fprintf(stderr, "Bad value after after 'pc=' option\n");
                    usage_old();
                    return SG_LIB_SYNTAX_ERROR;
                }
                optsp->do_pc = u;
            } else if (0 == strncmp(cp, "start", 5)) {
                if (startstop == 0)
                    ambigu = 1;
                else
                    startstop = 1;
            } else if (0 == strncmp(cp, "stop", 4)) {
                if (startstop == 1)
                    ambigu = 1;
                else
                    startstop = 0;
            } else if (0 == strncmp(cp, "old", 3))
                ;
            else if (jmp_out) {
                fprintf(stderr, "Unrecognized option: %s\n", cp);
                usage_old();
                return SG_LIB_SYNTAX_ERROR;
            }
        } else if (0 == strcmp("0", cp)) {
            if (1 == startstop)
                ambigu = 1;
            else
                startstop = 0;
        } else if (0 == strcmp("1", cp)) {
            if (0 == startstop)
                ambigu = 1;
            else
                startstop = 1;
        } else if (0 == optsp->device_name)
                optsp->device_name = cp;
        else {
            fprintf(stderr, "too many arguments, got: %s, not "
                    "expecting: %s\n", optsp->device_name, cp);
            usage_old();
            return SG_LIB_SYNTAX_ERROR;
        }
        if (ambigu) {
            fprintf(stderr, "please, only one of 0, 1, --eject, "
                    "--load, --start or --stop\n");
            usage_old();
            return SG_LIB_SYNTAX_ERROR;
        } else if (0 == startstop)
            ++optsp->do_stop;
        else if (1 == startstop)
            ++optsp->do_start;
    }
    return 0;
}

static int process_cl(struct opts_t * optsp, int argc, char * argv[])
{
    int res;
    char * cp;

    cp = getenv("SG3_UTILS_OLD_OPTS");
    if (cp) {
        optsp->opt_new = 0;
        res = process_cl_old(optsp, argc, argv);
        if ((0 == res) && optsp->opt_new)
            res = process_cl_new(optsp, argc, argv);
    } else {
        optsp->opt_new = 1;
        res = process_cl_new(optsp, argc, argv);
        if ((0 == res) && (0 == optsp->opt_new))
            res = process_cl_old(optsp, argc, argv);
    }
    return res;
}


int main(int argc, char * argv[])
{
    int fd, res;
    int ret = 0;
    struct opts_t opts;

    memset(&opts, 0, sizeof(opts));
    opts.do_fl = -1;    /* only when >= 0 set FL bit */
    res = process_cl(&opts, argc, argv);
    if (res)
        return SG_LIB_SYNTAX_ERROR;
    if (opts.do_help) {
        if (opts.opt_new)
            usage();
        else
            usage_old();
        return 0;
    }
    if (opts.do_version) {
        fprintf(stderr, "Version string: %s\n", version_str);
        return 0;
    }

    if (opts.do_start && opts.do_stop) {
        fprintf(stderr, "Ambiguous to give both '--start' and '--stop'\n");
        return SG_LIB_SYNTAX_ERROR;
    }
    if (opts.do_load && opts.do_eject) {
        fprintf(stderr, "Ambiguous to give both '--load' and '--eject'\n");
        return SG_LIB_SYNTAX_ERROR;
    }
    if (opts.do_load)
       opts.do_start = 1;
    else if ((opts.do_eject) || (opts.do_stop))
       opts.do_start = 0;
    else if (opts.opt_new && opts.do_loej && (0 == opts.do_start))
        opts.do_start = 1;      /* --loej alone in new interface is load */
    else if ((0 == opts.do_loej) && (-1 == opts.do_fl) && (0 == opts.do_pc))
       opts.do_start = 1;
    /* default action is to start when no other active options */
        
    if (0 == opts.device_name) {
        fprintf(stderr, "No DEVICE argument given\n");
        if (opts.opt_new)
            usage();
        else
            usage_old();
        return SG_LIB_SYNTAX_ERROR;
    }

    if (opts.do_fl >= 0) {
        if (opts.do_start == 0) {
            fprintf(stderr, "Giving '--fl=FL' with '--stop' (or "
                    "'--eject') is invalid\n");
            return SG_LIB_SYNTAX_ERROR;
        }
        if (opts.do_pc > 0) {
            fprintf(stderr, "Giving '--fl=FL' with '--pc=PC' "
                    "when PC is non-zero is invalid\n");
            return SG_LIB_SYNTAX_ERROR;
        }
    }
                
    fd = sg_cmds_open_device(opts.device_name, 0 /* rw */, opts.do_verbose);
    if (fd < 0) {
        fprintf(stderr, "Error trying to open %s: %s\n",
                opts.device_name, safe_strerror(-fd));
        return SG_LIB_FILE_ERROR;
    }

    res = 0;
    if (opts.do_fl >= 0)
        res = sg_ll_start_stop_unit(fd, opts.do_immed, opts.do_fl, 0 /* pc */,
                                    1 /* fl */, 1 /* loej */,
                                    1 /*start */, 1 /* noisy */,
                                    opts.do_verbose);
    else if (opts.do_pc > 0)
        res = sg_ll_start_stop_unit(fd, opts.do_immed, 0, opts.do_pc, 0, 0,
                                    0, 1, opts.do_verbose);
    else
        res = sg_ll_start_stop_unit(fd, opts.do_immed, 0, 0, 0, opts.do_loej,
                                    opts.do_start, 1, opts.do_verbose);
    ret = res;
    if (res) {
        if (opts.do_verbose < 2) {
            if (SG_LIB_CAT_INVALID_OP == res)
                fprintf(stderr, "command not supported\n");
            else if (SG_LIB_CAT_NOT_READY == res)
                fprintf(stderr, "device not ready\n");
            else if (SG_LIB_CAT_UNIT_ATTENTION == res)
                fprintf(stderr, "unit attention\n");
            else if (SG_LIB_CAT_ABORTED_COMMAND == res)
                fprintf(stderr, "aborted command\n");
            else if (SG_LIB_CAT_ILLEGAL_REQ == res)
                fprintf(stderr, "invalid field in cdb\n");
        }
        fprintf(stderr, "START STOP UNIT command failed\n");
    }
    res = sg_cmds_close_device(fd);
    if ((res < 0) && (0 == ret))
        return SG_LIB_FILE_ERROR;
    return (ret >= 0) ? ret : SG_LIB_CAT_OTHER;
}
